File: src/main/java/org/jukeboxmc/form/ButtonList.java
package org.jukeboxmc.form;

import com.nimbusds.jose.shaded.json.JSONArray;
import com.nimbusds.jose.shaded.json.JSONObject;
import org.jukeboxmc.form.element.Button;
import org.jukeboxmc.form.element.ImageButton;

import java.util.ArrayList;
import java.util.List;

/**
 * @author GoMint
 * @version 1.0
 */
public class ButtonList extends Form<String> {

    private final List<Button> buttons = new ArrayList<>();
    private String content = "";

    public ButtonList( String title ) {
        super( title );
    }

    public ButtonList setContent( String content ) {
        this.content = content;
        this.dirty = true;
        return this;
    }

    public ButtonList addButton( String id, String text ) {
        Button button = new Button( id, text );
        this.buttons.add( button );
        this.dirty = true;
        return this;
    }

    public ButtonList addImageButton( String id, String text, String imagePath ) {
        ImageButton imageButton = new ImageButton( id, text, imagePath );
        this.buttons.add( imageButton );
        this.dirty = true;
        return this;
    }

    @Override
    public String getFormType() {
        return "form";
    }

    @Override
    public JSONObject toJSON() {
        // Fast out when cached
        if ( this.cache != null && !this.dirty ) {
            return this.cache;
        }

        // Create new JSON view of this form
        JSONObject jsonObject = super.toJSON();
        JSONArray content = new JSONArray();

        for ( Button button : this.buttons ) {
            content.add( button.toJSON() );
        }

        jsonObject.put( "content", this.content );
        jsonObject.put( "buttons", content );

        // Cache and return
        this.cache = jsonObject;
        this.dirty = false;
        return this.cache;
    }

    @Override
    public String parseResponse( String json ) {
        // Input is LF terminated
        try {
            int buttonId = Integer.parseInt( json.trim() );
            if ( buttonId + 1 > this.buttons.size() ) {
                return null;
            }

            return this.buttons.get( buttonId ).getId();
        } catch ( NumberFormatException e ) {
            return null;
        }
    }

}

File: JukeboxMC-API/src/main/kotlin/org/jukeboxmc/api/form/ButtonList.kt
package org.jukeboxmc.api.form

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import org.jukeboxmc.api.form.element.Button
import org.jukeboxmc.api.form.element.ImageButton

/**
 * @author Kaooot
 * @version 1.0
 */
class ButtonList(title: String) : Form<String>(title) {

    private val buttons = arrayListOf<Button>()
    private var content: String = ""

    override fun getFormType(): String = "form"

    override fun parseResponse(json: String): String? {
        val buttonId = json.trim().toInt()
        if (buttonId + 1 > this.buttons.size) {
            return null
        }
        return this.buttons[buttonId].getId()
    }

    override fun toJson(): JsonObject {
        if (this.cache != null && !this.dirty) {
            return this.cache!!
        }
        val jsonObject = super.toJson()
        val buttons = JsonArray()
        for (button in this.buttons) {
            buttons.add(button.toJSON())
        }
        jsonObject.addProperty("content", this.content)
        jsonObject.add("buttons", buttons)
        this.cache = jsonObject
        this.dirty = false
        return this.cache!!
    }

    fun setContent(content: String): ButtonList {
        this.content = content
        this.dirty = true
        return this
    }

    fun addButton(id: String, text: String): ButtonList {
        val button = Button(id, text)
        this.buttons.add(button)
        this.dirty = true
        return this
    }

    fun addImageButton(id: String, text: String, image: String): ButtonList {
        val imageButton = ImageButton(id, text, image)
        this.buttons.add(imageButton)
        this.dirty = true
        return this
    }
}